### ISSUE TITLE
Can we remove runtime types from Chapel?

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#19292]
The runtime types of arrays and domains in Chapel present both semantic complications and implementation difficulties. One example is #11549.

This issue proposes to investigate what it would take to switch array and domain types to purely-static types. For example:

* What aspects of the language definition would need to change? 
* In what ways do user codes rely on array/domain types having a runtime component?

**Comment by damianmoz** [comment]
Any chance of more examples please?

Is there a way to define a record which contains an array of **real**(w)?

**Comment by mppf** [comment]
@damianmoz - yes I can provide some examples about what runtime types do today.

``` chapel


config const n: int;
var B: [1..n] int; // has domain {1..n}

// for an array, the runtime type is the domain
// for a domain, the runtime type is the distribution

// So, B's runtime type is a tuple like this: ({1..n}, )

// We can declare another array with the same shape using the runtime type:
var C: B.type;

// We can also do so by calling a function that accepts a type
var D = f(B.type);

proc f(type t) {
  var a: t;
  return a;
}


// What does the above 'var a: t' do for an array type t?
// Suppose we did not have runtime types.
// Then t would be [??who knows] int
//  (no information about the domain/bounds/size)
// So, what could we do?
//  * error, can't default initialize an array
//  * make an array of 0 elements
```

A problem with runtime types is that they do not generalize to records containing arrays:

``` chapel
record GR { // "Generic Record"
  type t;
}

type MyGR = GR(B.type);
var x: MyGR.t;
// A reasonable user might expect x to have the same bounds as C
// But this does not work today (probably failing with a weird error at runtime)
```

> Is there a way to define a record which contains an array of real(w)?

Certainly, and we do this quite frequently. Here is a common pattern in a form that compiles and runs:

``` chapel
record R {
  var D; // indicates the domain (bounds) of the array
  var array: [D] int;
}
var x = new R({1..10});
writeln(x); // outputs (D = {1..10}, array = 0 0 0 0 0 0 0 0 0 0)
```

Note that this pattern does not use runtime types.

**Comment by vasslitvinov** [comment]
One aspect: how to write, in Chapel, the non-runtime type of a domain/array with a non-default domain map? We want to ensure it is doable. My first guess: `domain(1) dmapped Block` would be generic over all Block instances.

**Comment by vasslitvinov** [comment]
Noting that the code in the OP of #20050 will work as-is if we remove runtime types. This is because it does not involve default-initializing arrays other than with known domains.

**Comment by damianmoz** [comment]
@mppf, you wrote
> Note that this pattern does not use runtime types.

I was thinking more of an example with an array `X` where the bounds opf the array are not known at compile time and whose mathematical base type is **real**(_w_) where _w_ might not be a **param**, although in practice it is most likely a **param** or a **config** **param**.

**Comment by damianmoz** [comment]
The mathematical concept of a base type is what Chapel calls _eltype_.

I try and write code that works as well for say a 32 bit real as a 64 bit real and hopefully even 128-bit reals. Sadly, I have generic code **real**(_w_) code which may fail for some extreme values of **real**(_16_).

**Comment by vasslitvinov** [comment]
@damianmoz how do you see runtime types -  or other existing or potential features - helping you in this case?

**Comment by mppf** [comment]
@damianmoz 

> @mppf, you wrote
> 
> > Note that this pattern does not use runtime types.
> 
> I was thinking more of an example with an array `X` where the bounds opf the array are not known at compile time and whose mathematical base type is **real**(_w_) where _w_ might not be a **param**, although in practice it is most likely a **param** or a **config** **param**.

Yes. I think there is a terminology issue here. Supposing `n` is a `config const` (or any other runtime-variable value). Then if we write `var X: [1..n] int`, the syntax says that the type is `[1..n] int` which indeed is only fully known until runtime. However in the proposal to "remove runtime types", I don't think we are talking about changing this syntax. So the syntax would still mean whatever it means today: namely, to create an array with `n` elements starting with element `1`. `n` could continue to vary at runtime.

Basically, I think the boiled-down form of proposal here would be to remove support for keeping track of the array bounds in any `type` variable that refers to an array type or in `someArray.type` expressions. https://github.com/chapel-lang/chapel/issues/19292#issuecomment-1048876014 has examples of these patterns specifically.

**Comment by damianmoz** [comment]
Give me some time to ponder this. But something about this proposal seems overly restrictive.

If you allow a variable to have a runtime type, why cannot you define a **type** variable to have that same type. Does not that imply some serious restrictions on the ability to use this in a generic case.

**Comment by vasslitvinov** [comment]
@damianmoz -

> If you allow a variable to have a runtime type, why cannot you define a type variable to have that same type. Does not that imply some serious restrictions on the ability to use this in a generic case?

Indeed, a prominent feature of this proposal is to disallow `var A: t;` where `t` is an array type. We may make exceptions for the cases where the domain of `t` is obvious to the compiler. `t` could be defined in a generic context, such as a type argument of a generic function.

This will go hand-in-hand with restrictions on uses of the runtime type of an array variable.

How well am I understanding your question?

**Comment by damianmoz** [comment]
I am getting lost.

Assuming that `m` and `n` are both a **const**, what is the difference between the two cases:
```
// Case A
var a : [m..n, m..n] real(32);

// Case B
type myMat = [m..n, m..n] real(32);
var a : myMat
```

How would I declare a record 'R' containing an **array** of dimensions `n * n` where the earlier assumptions hold and type **real**(_w_) where `w` is a **param**?

**Comment by damianmoz** [comment]
I note the example from @mppf but change it slightly where `n` is a **const**.  Does this work and is it not a run-time type?
```chapel
record R
{
  param w;
  const n;
  var array: [1..n] real(w);
}
var x = new R(32, n);
writeln(x);
```
I never understand why this needs a **new**. I always get worried I will have to **delete** it explicitly.

I note that I cannot declare a type inside that record and use it in the definition of `a` within `R`.

**Comment by bradcray** [comment]
> I never understand why this needs a new. I always get worried I will have to delete it explicitly.

I think the short answer to this question is that Chapel does not distinguish, syntactically, between types and values as, say, C++ does.  So if the `new` were not required to invoke a record's initializer, and you could just write `R(32, n)` then it would be ambiguous whether you were attempting to specify a type expression (e.g., `myRealArray(2)` might mean "my 2D array of reals)  or a value initializer (e.g., it could also mean "construct me a 2-element array of reals" or "construct me an array of reals initialzed to 2".

So the `new` in Chapel is a way of prefixing an expression that might otherwise look like a type and interpret it as a value.  We've considered extending this to all types so you could say `new int(42)` or `new real(3.14)` for example, to support generic programming ("take a `type t` and call `new t(args)`) but have not done so yet.

All that said, I'd also suggest getting past thinking of `new` as requiring `delete` since even most class types in Chapel don't require `delete`s to match their `new`s—only the `unmanaged` class type.

I'll leave the other questions to those who are advocating for this change other than to note that `const` is still an execution-time property in Chapel, so `n` would need to be a `param` to make it completely able to be reasoned about statically, and that would require support for `param` ranges and domains (which we've discussed being desirable but not yet implemented on other issues).

**Comment by damianmoz** [comment]
Is this run-time typing and what you want to eliminate?
```chapel
record Stuff
{
  type R;
  const n;
  var array: [1..n] R;
}
var x = new Stuff(real(32), n);
writeln(x);
```

**Comment by vasslitvinov** [comment]
Hi Damian, thank you for the discussion. Be assured that `real` types, all numeric types, strings, ranges, and more will continue working the same way under this proposal. This is because only arrays and domains have "runtime types" in Chapel at present.

Your example with `record Stuff` should continue working like it does today. This is because the only use of a runtime type, which is the array type`[1..n] R`, is such that its domain is easily accessible by the compiler. 

In your earlier examples:
```chpl
// Case A
var a : [m..n, m..n] real(32);

// Case B
type myMat = [m..n, m..n] real(32);
var a : myMat
```

Case A falls in the same "easily accessible" category. Case B is a bit borderline, we may decide either to handle it like Case A or require the user to rewrite it to something like Case A.

This proposal will likely disallow this variation on your `record Stuff`:
```chpl
record Stuff
{
  type R;
  const n;
  var array: R;
}
var x = new Stuff([1..n] real(32), n);
```
How much impact do you think this restriction will have on your Chapel development?

**Comment by damianmoz** [comment]
That final one, none as far as I can tell.

If Case B is not allowed, it might just be a nuisance.

**Comment by mppf** [comment]
One thing occurs to me where runtime-types might actually be counter-productive.

I'm thinking about how the way that a sparse array's runtime type is connected to the nonzero elements might trip up a programmer.

One way this might come up is with a function that returns an array resulting from adding in 1 on the diagonal. Here is what I would think would be the "obvious" code for that:

``` chapel
proc addOnesDiagonal(A) {
  var B = A;
  var n = ... // compute n from A's domain
  for i in 1..n {
    B[i, i] += 1.0;
  }
}
```

But, this code would not work. I would expect it to give out-of-bounds errors when `B[i,i] += 1.0` is run for an `i` where `(i,i)` does not yet exist in `A.domain`. Instead, the sparse array needs a separate action to modify the set of nonzero elements (separate from just assigning to the array). Additionally, this set of nonzero elements is part of the array's type and modifying it will change other arrays with the same type. I think that this situation with sparse arrays is connected to runtime types in Chapel. I am far from being an expert in sparse arrays but the situation seems cumbersome to me.

Relatedly, declaring the signature like this is also wrong:
``` chapel
proc addOnesDiagonal(A): A.type { ... }
```

**Comment by bradcray** [comment]
> Instead, the sparse array needs a separate action to modify the set of nonzero elements (separate from just assigning to the array). Additionally, this set of nonzero elements is part of the array's type and modifying it will change other arrays with the same type.

To me, this seems similar to the tradeoffs between associative domains and arrays vs. maps in Chapel.  The former can be annoying because you have to modify the domain before storing into the array, but has the advantage of supporting multiple arrays with the same domain and amortizing those overheads (vs. having multiple maps that redundantly store the same keys).  Sparse arrays also require separate modification of their domains to change their allocation, but support similar amortization, where a sparse matrix library type (say) might provide better ergonomics, but with the downside of no ability to share the sparsity structure.

**Comment by vasslitvinov** [comment]
@bradcray what do you think about replacing associative arrays with a separate map datatype where a map uses an external key set?

Or perhaps not a separate datatype, rather an "external keys" field for maps that can be either "nothing" (today's maps) or a reference to a key set?

**Comment by vasslitvinov** [comment]
I added a couple of challenging scenarios to the OP.

**Comment by mppf** [comment]
@bradcray 

> but with the downside of no ability to share the sparsity structure.

Yes, exactly. What I am trying to ask is this: do we know of real use cases or applications for sparse arrays where sharing the sparsity structure between arrays is important or useful? As I said, I'm not an expert in sparse computations, but it seems to not be what I would want in the cases that I can think of.

If it turns out to be not-that-useful, we could consider having sparse arrays connected only to the bounding box rectangular domain (so "resizing" would allow read or write array accesses to the new elements, which would start out as the "zero" value). IMO this is equally reasonable to the design that we have, in terms of how to generalize dense array behavior to sparse arrays. Or, we could change the sparse array support to no longer be the "array" type and rather something else (including, a distributed implementation), similarly to how we added `map`.

I also don't have at-the-ready an idea of cases where having multiple arrays sharing a domain is valuable in other cases, including associate array and dense array case. I am imagining that such use cases would not be related to the use cases for sparse arrays, but perhaps I am wrong about that.

**Comment by bradcray** [comment]
> what do you think about replacing associative arrays with a separate map datatype where a map uses an external key set?

I can see why that could be attractive (except that it complicates the `map` implementation), but I guess I don't see why that's better / easier / preferable to associative domains and arrays.

> do we know of real use cases or applications for sparse arrays where sharing the sparsity structure between arrays is important or useful?

I've admittedly only used them in modest-sized benchmarks rather than real applications, but based on those uses, it's not difficult for me to imagine sharing sparse domain across multiple sparse arrays—particularly when using them to represent a subset of elements in a spatial domain.

> I also don't have at-the-ready an idea of cases where having multiple arrays sharing a domain is valuable in other cases, including associate array and dense array case.

I think Engin's locality-based compiler optimizations rely on reasoning about the relationship between domains and arrays  (which was definitely one of the original motivations for raising them to a language concept distinct from arrays).

**Comment by vasslitvinov** [comment]
> >   what do you think about replacing associative arrays with a separate map datatype where a map uses an external key set?
> 
> I can see why that could be attractive (except that it complicates the map implementation), but I guess I don't see why that's better / easier / preferable to associative domains and arrays.

The advantage would be to reduce the number of things built into the language.

Also we can look at it in two ways:

* Chapel's design enables additional compile-time reasoning and optimizations due to the built-in relationship between arrays and domains.
* User-defined data structures cannot get some of the benefits that are available only to arrays.

Thinking out loud. This topic is not on my front burner.

**Comment by vasslitvinov** [comment]
I am removing the 2.0 label as we are retaining the runtime types for now.
